lib/extensions.js

```diff
@@ -80,37 +80,43 @@ module.exports = function(realmConstructor, context) {
     const { DefaultNetworkTransport } = require('realm-network-transport');
     realmConstructor._networkTransport = new DefaultNetworkTransport();
 
-    Object.defineProperty(realmConstructor.Object.prototype, "toJSON", {
-        value: function () {
-            const result = {}
-            Object.keys(this).forEach(p => result[p] = this[p]);
-            Object.keys(Object.getPrototypeOf(this)).forEach(p => result[p] = this[p]);
-            return result;
+    Object.defineProperty(realmConstructor.Collection.prototype, "toJSON", {
+        value: function (_, cache = new Map()) {
+            return this.map((item, index) => item.toJSON ? item.toJSON(index.toString(), cache) : item);
         },
 
         writable: true,
         configurable: true,
         enumerable: false
     });
 
-    Object.defineProperty(realmConstructor.Object.prototype, "keys", {
-        value: function () {
-            return Object.keys(this).concat(Object.keys(Object.getPrototypeOf(this)));
-        },
+    Object.defineProperty(realmConstructor.Object.prototype, "toJSON", {
+        value: function (_, cache = new Map()) {
+            // Construct a reference-id of table-name & objectId
+            const id = this.constructor.name + this._objectId();
```

Comment:
we are using the `objectId` but in the generated json we call it `$refId` and `$ref`. Doesn't it make sense to use the same naming like `$obj` and '$objId'. Also why not name this with underscore and full property name `_objectId` and `_object`. It will be more consistent with the actual `_objectId` every Realm Object has.

Reply:
`$refId` is a combination of the object-type/table-name & `objectId`, because the `objectId` apparently only is unique with in the table (see @kneth's comment here: https://github.com/realm/realm-js/pull/3013).
It's is only included in the JSON serialization to indicate where circular references already exists. By not only consisting of `objectId` I named it to something matching the current context (JSON serialization).

A property with a prefixed underscore indicates a private property (in JS), which should (most likely) not be serialized.
In Realm `_objectId` is a function, which are not serializable, so I would argue that `_objectId` showing up (in an extended state) as a string, would be more confusing.

Reply:
sure thanks.